oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java

```diff
@@ -83,56 +85,112 @@
-               if (appliance.minHddSize() > volume.getSize()) {
-                  minHddSize = appliance.minHddSize();
-               }
-            }
-            Hdd.CreateHdd hdd = Hdd.CreateHdd.create(minHddSize, volume.isBootDevice());
-            hdds.add(hdd);
-         } catch (Exception ex) {
-            throw Throwables.propagate(ex);
-
+      String imageId = image.getId();
+      HardwareFlavour hardwareModel = isFlavor(hardware.getId());
+      boolean isBaremetal = hardware.getName() == null ? false : hardware.getName().contains(baremetalModelsKey);
+      ServerAppliance workingImage = null;
+
+      //choose the correct image based on the server type baremetal or cloud
+      if (hardwareModel != null) {
+         if (isBaremetal) {
+            workingImage = findWorkingImage(image.getId(), Types.ServerTypeCompatibility.baremetal);
+         } else {
+            workingImage = findWorkingImage(image.getId(), Types.ServerTypeCompatibility.cloud);
+         }
+         if (workingImage != null) {
+            imageId = workingImage.id();
+         }
+      } else {
+         //check if image is of Type IMAGE
+         workingImage = findWorkingImage(image.getId(), Types.ServerTypeCompatibility.cloud);
+         if (workingImage != null) {
+            imageId = workingImage.id();
          }
       }
 
-      // provision server
-      Server server = null;
-      Double cores = ComputeServiceUtils.getCores(hardware);
-      Double ram = (double) hardware.getRam();
-      if (ram < 1024) {
-         ram = 0.5;
-      } else {
-         ram = ram / 1024;
+      //configuring Firewall rules
+      Map<Integer, Integer> portsRange = getPortRangesFromList(inboundPorts);
+      List<FirewallPolicy.Rule.CreatePayload> rules = new ArrayList<FirewallPolicy.Rule.CreatePayload>();
+
+      for (Map.Entry<Integer, Integer> range : portsRange.entrySet()) {
+         FirewallPolicy.Rule.CreatePayload rule = FirewallPolicy.Rule.CreatePayload.builder()
+                 .portFrom(range.getKey())
+                 .portTo(range.getValue())
+                 .protocol(Types.RuleProtocol.TCP)
+                 .build();
+         rules.add(rule);
+      }
+      String firewallPolicyId = "";
+      if (inboundPorts.length > 0) {
+         FirewallPolicy firewallPolicy = api.firewallPolicyApi().create(FirewallPolicy.CreateFirewallPolicy.create(name + " firewall policy", "desc", rules));
+         firewallPolicyId = firewallPolicy.id();
       }
 
+      //
+      // provision server
+      Server server = null;
+      org.apache.jclouds.oneandone.rest.domain.Hardware.CreateHardware hardwareRequest = null;
       try {
-         org.apache.jclouds.oneandone.rest.domain.Hardware.CreateHardware hardwareRequest
-                 = org.apache.jclouds.oneandone.rest.domain.Hardware.CreateHardware.create(cores, 1, ram, hdds);
+         if (hardwareModel != null) {
```

Comment:
this logic to build the hardwareRequest is quite complex. would it make sense to extract it in its own method, and add some unit tests, ideally

Reply:
done [here](https://github.com/StackPointCloud/jclouds-labs/blob/oneandone-baremetal-update/oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/strategy/GenerateHardwareRequest.java)